Zend/zend_object_handlers.c

```diff
@@ -175,6 +176,134 @@ ZEND_API HashTable *zend_std_get_debug_info(zend_object *object, int *is_temp) /
+	zend_fcall_info fci;
+	zend_fcall_info_cache fcic;
+	zval params[2] = {*op1, *op2};
+
+	EG(fake_scope) = NULL;
+
+	/* op handlers like __add are called with two operands op1, op2 */
+	fci.size = sizeof(fci);
+	fci.retval = result;
+	fci.param_count = 2;
+	fci.params = params;
+	fci.no_separation = 1;
+	//ZVAL_UNDEF(&fci.function_name); /* Unused */
+
+	do
+	{
+		fci.object = zobj;
+
+		/* Determine what handler should be used, based on opcode */
+		switch (opcode)
+		{
+		case ZEND_ADD:
+			fcic.function_handler = ce->__add;
+			ZVAL_STRING(&fci.function_name, ZEND_ADD_FUNC_NAME);
+			break;
+		case ZEND_SUB:
+			fcic.function_handler = ce->__sub;
+			ZVAL_STRING(&fci.function_name, ZEND_SUB_FUNC_NAME);
+			break;
+		case ZEND_MUL:
+			fcic.function_handler = ce->__mul;
+			ZVAL_STRING(&fci.function_name, ZEND_MUL_FUNC_NAME);
+			break;
+		case ZEND_DIV:
+			fcic.function_handler = ce->__div;
+			ZVAL_STRING(&fci.function_name, ZEND_DIV_FUNC_NAME);
+			break;
+		case ZEND_POW:
+			fcic.function_handler = ce->__pow;
+			ZVAL_STRING(&fci.function_name, ZEND_POW_FUNC_NAME);
+			break;
+		case ZEND_MOD:
+			fcic.function_handler = ce->__mod;
+			ZVAL_STRING(&fci.function_name, ZEND_MOD_FUNC_NAME);
+			break;
+		case ZEND_CONCAT:
+			fcic.function_handler = ce->__concat;
+			ZVAL_STRING(&fci.function_name, ZEND_CONCAT_FUNC_NAME);
+			break;
+		case ZEND_SL:
+			fcic.function_handler = ce->__sl;
+			ZVAL_STRING(&fci.function_name, ZEND_SHIFT_LEFT_FUNC_NAME);
+			break;
+		case ZEND_SR:
+			fcic.function_handler = ce->__sr;
+			ZVAL_STRING(&fci.function_name, ZEND_SHIFT_RIGHT_FUNC_NAME);
+			break;
+		case ZEND_BW_OR:
+			fcic.function_handler = ce->__or;
+			ZVAL_STRING(&fci.function_name, ZEND_OR_FUNC_NAME);
+			break;
+		case ZEND_BW_AND:
+			fcic.function_handler = ce->__and;
+			ZVAL_STRING(&fci.function_name, ZEND_AND_FUNC_NAME);
+			break;
+		case ZEND_BW_XOR:
+			fcic.function_handler = ce->__xor;
+			ZVAL_STRING(&fci.function_name, ZEND_XOR_FUNC_NAME);
+			break;
+		default:
+			return FAILURE;
+			break;
+		}
+
+		/* Check if function exists, check on other operand if possible */
+		if (fcic.function_handler == NULL)
+		{
+			if(zobj == Z_OBJ_P(op1) && Z_TYPE_P(op2) == IS_OBJECT) {
+				zobj = Z_OBJ_P(op2);
+				ce = Z_OBJCE_P(op2);
+				zval_ptr_dtor(&fci.function_name);
+				/* Retry checking if other operand has method */
+				continue;
+			}
+
+			zend_error(E_NOTICE, "You have to implement the %s function in class %s to use this operator with an object!",
+					   Z_STRVAL(fci.function_name), ZSTR_VAL(ce->name));
+			zval_ptr_dtor(&fci.function_name);
+			return FAILURE;
+		}
+
+		fcic.called_scope = ce;
+		fcic.object = zobj;
+
+		int tmp = zend_call_function(&fci, &fcic);
+
+		if (!EG(exception) && (Z_TYPE_P(result) == IS_UNDEF || Z_TYPE_P(result) == IS_NULL))
+		{
+			zend_error(E_ERROR, "Method %s::%s must return a non-null value", ZSTR_VAL(ce->name), Z_STRVAL(fci.function_name));
```

Comment:
This should use `zend_type_error()`.

Reply:
Though I'm wondering: Might it make sense to allow a null return value to indicate that this operand combination is not supported? In particular, I'm thinking that if you have `$a * $b` and `A::__mul($a, $b)` returns null, then `B::__mul($a, $b)` would be tried.

So if some library that provides A implements __mul but does not know about B, then B still has a chance to implement a __mul behavior if it knows about A.

Reply:
> This should use `zend_type_error()`.

Thank you for the hint.

> Might it make sense to allow a null return value to indicate that this operand combination is not supported?

I thought about this too. I think some mechanisms like this is highly useful, but i wonder if returning null value is the most intuitive way for that (returning null looks like the operator would return null). Maybe introducing a special Exception/Throwable, like OperandsNotSupportedException, that will be thrown in that cases and signal to use the other operands handler.

Reply:
Throwing does carry overhead, perhaps a better option is a separate ce that can only be constructed via a call to an internal constant or method?

It could then return that, instead of throwing.

`
class Foo { 
public static function __add(Foo $x, $y): Foo|ArithmeticOperationUnsupported  { 
   if (!some_condition($y)) { 
      return Arithmetic::OPERATION_UNSUPPORTED;
   }
 
   return new Foo(...); 
}
}
`